fdbserver/GrvProxyTransactionTagThrottler.actor.cpp

```diff
@@ -0,0 +1,329 @@
+	req.proxyTagThrottledDuration = now() - startTime;
+}
+
+void GrvProxyTransactionTagThrottler::TagQueue::setRate(double rate) {
+	if (rateInfo.present()) {
+		rateInfo.get().setRate(rate);
+	} else {
+		rateInfo = GrvTransactionRateInfo(rate);
+	}
+}
+
+void GrvProxyTransactionTagThrottler::updateRates(TransactionTagMap<double> const& newRates) {
+	for (const auto& [tag, rate] : newRates) {
+		auto it = queues.find(tag);
+		if (it == queues.end()) {
+			queues[tag] = TagQueue(rate);
+		} else {
+			it->second.setRate(rate);
+		}
+	}
+
+	// Clean up tags that did not appear in newRates
+	for (auto& [tag, queue] : queues) {
+		if (newRates.find(tag) == newRates.end()) {
+			queue.rateInfo.reset();
+		}
+	}
+
+	// TODO: Use std::erase_if in C++20
+	for (auto it = queues.begin(); it != queues.end();) {
+		const auto& [tag, queue] = *it;
+		if (queue.requests.empty() && !queue.rateInfo.present()) {
+			it = queues.erase(it);
+		} else {
+			++it;
+		}
+	}
+}
+
+void GrvProxyTransactionTagThrottler::addRequest(GetReadVersionRequest const& req) {
+	ASSERT(req.isTagged());
+	auto const& tag = req.tags.begin()->first;
+	if (req.tags.size() > 1) {
+		// The GrvProxyTransactionTagThrottler assumes that each GetReadVersionRequest
+		// has at most one tag. If a transaction uses multiple tags and
+		// SERVER_KNOBS->ENFORCE_TAG_THROTTLING_ON_PROXIES is enabled, there may be
+		// unexpected behaviour, because only one tag is used for throttling.
+		TraceEvent(SevWarnAlways, "GrvProxyTransactionTagThrottler_MultipleTags")
+		    .detail("NumTags", req.tags.size())
+		    .detail("UsingTag", printable(tag));
+	}
+	queues[tag].requests.emplace_back(req);
+}
+
+void GrvProxyTransactionTagThrottler::releaseTransactions(double elapsed,
+                                                          SpannedDeque<GetReadVersionRequest>& outBatchPriority,
+                                                          SpannedDeque<GetReadVersionRequest>& outDefaultPriority) {
+	struct TagInfo {
+		// Store pointers here to avoid frequent std::unordered_map lookups
+		TagQueue* queue;
+		uint32_t* numReleased;
+		// Sequence number of the first queued request
+		int64_t nextSeqNo;
+		bool operator<(TagInfo const& rhs) const { return nextSeqNo < rhs.nextSeqNo; }
+		explicit TagInfo(TagQueue& queue, uint32_t& numReleased) : queue(&queue), numReleased(&numReleased) {
+			ASSERT(!this->queue->requests.empty());
+			nextSeqNo = this->queue->requests.front().sequenceNumber;
+		}
+	};
+
+	// Track transactions released for each tag
+	std::vector<std::pair<TransactionTag, uint32_t>> transactionsReleased;
+	transactionsReleased.reserve(queues.size());
+
+	std::priority_queue<TagInfo> pq;
+	for (auto& [tag, queue] : queues) {
+		if (queue.rateInfo.present()) {
+			queue.rateInfo.get().startReleaseWindow();
+		}
+		if (!queue.requests.empty()) {
+			auto& [_, count] = transactionsReleased.emplace_back(tag, 0);
+			pq.emplace(queue, count);
```

Comment:
There's a lot going on here which I would've missed if I didn't stare at this for multiple minutes. Basically you're lazily adding a counter to `transactionsReleased` but this counter is then modified by the object in `pq`...

I am not sure whether this is the best way of doing this (it might be -- I understand this code is performance critical). But please add comments to this code block